### ISSUE TITLE
Alt text of maths fallback images is missing in wikipedia_en_maths_novid_2017-11.zim

### ISSUE THREAD

**Issue by Jaifroid** [openzim/mwoffliner#164]
This issue is similar to #114 (mwoffliner should store the alt text of maths fallback images). #114 was closed in September because the latest Wikipedia novid ZIMs were correctly storing the maths alt text.

However, with the file `wikipedia_en_maths_novid_2017-11.zim` this bug has re-appeared. This file only has the SVG images for maths equations, and does not have the alt text which enables MathJax based solutions to display maths equations quickly and with low memory usage. Kiwix-JS-Windows uses a MathJax solution, because extracting SVGs is extremely memory hungry in the JavaScript app (indeed, it crashes the app on lower-memory systems).

Therefore it is very important for mwoffliner to retain the alt-text of maths equations.

**Comment by kelson42** [comment]
@Jaifroid I'm sorry you have to reopen that ticket... We will try to have a look quickly. Thx for the report.

**Comment by kelson42** [comment]
The problem is specific with the mobile layout. The problem is only fixed with the normal layout.

**Comment by kelson42** [comment]
@subbuss It seems the Parsoid does not deliver any "alt" attribute!? For example

```
<img src=\"https://wikimedia.org/api/rest_v1/media/math/render/svg/0d1ecb613aa2984f0576f70f86650b7c2a132538\" class=\"mwe-math-fallback-image-inline\" aria-hidden=\"true\" style=\"vertical-align: -0.338ex; width:1.059ex; height:1.676ex;\">
```

for the example in the "Wilson–Cowan_model" WPEN article. Is that on purpose?

**Comment by Jaifroid** [comment]
Just to add that another good article for testing in `wikipedia_en_medicine_novid_2018-01` is "Integral".

**Comment by Jaifroid** [comment]
Sorry to come back to this issue, but I notice a discrepancy between two recent ZIMs which might be pertinent in solving this:

`wikipedia_en_all_novid_2018-06.zim`, from June, *does* contain the TeX alt-text of maths equations, and all articles such as 'Sine', 'Wilson-Cowan Model' or 'Trigonometric Functions' render the multiple equations correctly in MathJax.

However, `wikipedia_en_medicine_novid_2018-08.zim`, from August, *does not* contain the TeX alt-text of maths equations, and so clients are forced to extract multiple SVGs for pages such as 'Wilson-Cowan Model'. Also, and more seriously, `wikipedia_en_maths_novid_2018-06.zim` *does not* contain the TeX either.

What is the difference in the way these ZIMs are produced? Maybe that can provide a clue as to how to address this issue. It really does make a huge difference on slow clients to be able to interpret the TeX (and we are forced to do this anyway for Stackexchange ZIMs).

**Comment by kelson42** [comment]
@Jaifroid I do not remember to have change anything related to that recently in mwoffliner... Like written before, this only impact ZIM files with the "mobile layout".

**Comment by ISNIT0** [comment]
Created an issue here: https://phabricator.wikimedia.org/T209277

**Comment by Jaifroid** [comment]
This issue is still persisting with the very latest ZIM formats, e.g. `wikipedia_es_all_novid_2019-06.zim`. On es.wikipedia.com, the alt text with the formulas, e.g. https://es.wikipedia.org/wiki/Seno_(trigonometría) , is present, but in our ZIM, the alt text has been removed, and the client is forced to extract SVGs instead of typesetting with MathJax. Taking a look at the Phabricator issue created by @ISNIT0, it seems to have been ignored. It is marked as "Needs Triage".

**Comment by kelson42** [comment]
@Jaifroid Have we a concrete problem somewhere with SVG?

**Comment by Jaifroid** [comment]
@kelson42 The issue has always been that on slow clients / old browsers on Kiwix JS / Kiwix JS Windows, it is extremely CPU-intensive to decompress SVGs (they are highly compressed), and pages with lots of them such as "Trigonometric Functions" are completely unusable on such devices. Hence, I implemented typesetting equations with MathJax instead, and it makes these pages perfectly usable on old/slow devices. However, that's not possible when the alt text, which contains the TeX source, is missing.

**Comment by kelson42** [comment]
@ISNIT0 How much work would that be get (back) png variant of this mathematical stuff?

**Comment by Jaifroid** [comment]
Yes, there are quite a few little SVG flags, etc., on Wikipedia pages. They are stored very efficiently in SVG format in the ZIM (take very little space), so there would be a trade-off between the CPU needed to decompress them as SVGs, and the extra disk space needed (?) to store them in a binary image format like PNG. PNGs are certainly retrieved much faster than SVGs, because the former only need to be extracted, not decompressed.

But it would still be excellent if we could get the `math/tex` strings. Wikipedia provides them, it's just a matter of a Parsoid dev writing probably one extra line of code. And anyway, it's good a11y to have alt text for images.

**Comment by kelson42** [comment]
@Jaifroid There is trade-off. I always thought it was in favour of SVG, but I might be wrong. Therefore, I try to re-evaluate the situation, once again. The other reason to do that, is that I don't believe https://phabricator.wikimedia.org/T209277 is going to be fixed soon.

**Comment by Jaifroid** [comment]
I also found a closed issue on Phabricator https://phabricator.wikimedia.org/T136915 from 2016 that claimed to have fixed the math/tex alt text issue. So I reopened that issue with a cross link to our issue. Let's see if it grabs anyone's attention...

**Comment by Jaifroid** [comment]
The Parsoid issue underlying this seems to be fixed. Looking a the output of https://en.wikipedia.org/api/rest_v1/page/mobile-sections/Wilson–Cowan_model, we see, for example:
```
<span class=\"mwe-math-element\">
    <span class=\"mwe-math-mathml-inline mwe-math-mathml-a11y\" style=\"display: none;\">
        <math xmlns=\"http://www.w3.org/1998/Math/MathML\" 
            alttext=\"{\\displaystyle 1-\\int _{t-r}^{t}E(t')dt'}\">
```
The alttext with the MathML is clearly visible in the last line above. I think it's just a question of including it in mwoffliner output (please!).

**Comment by kelson42** [comment]
@Jaifroid A miracle! Do you mean they might have fixed the bug without closing the Phabricator ticket?

**Comment by Jaifroid** [comment]
I don't know. I don't really know enough about Parsoid. Is that API a Parsoid API for mobile view? - https://en.wikipedia.org/api/rest_v1/page/mobile-sections/Wilson%E2%80%93Cowan_model . If so, it's fixed. If I've misunderstood, and that API just shows us the stringified HTML of the page, then probably it's not. I just noticed that the alttext is given by that API. Can you confirm?

**Comment by kelson42** [comment]
@Jaifroid Currently MWoffliner removes the whole DOM node with CSS class 'mwe-math-mathml-inline'. Would that help if we just keep that node?

It would give us something like this?
```
<span class="mwe-math-element"><span class="mwe-math-mathml-inline mwe-math-mathml-a11y" style="display: none;"><math xmlns="http://www.w3.org/1998/Math/MathML" alttext="{\displaystyle \int _{t-r}^{t}E(t')dt'}">
  <semantics>
    <mrow class="MJX-TeXAtom-ORD">
      <mstyle displaystyle="true" scriptlevel="0">
        <msubsup>
          <mo>∫</mo>
          <mrow class="MJX-TeXAtom-ORD">
            <mi>t</mi>
            <mo>−</mo>
            <mi>r</mi>
          </mrow>
          <mrow class="MJX-TeXAtom-ORD">
            <mi>t</mi>
          </mrow>
        </msubsup>
        <mi>E</mi>
        <mo stretchy="false">(</mo>
        <msup>
          <mi>t</mi>
          <mo>′</mo>
        </msup>
        <mo stretchy="false">)</mo>
        <mi>d</mi>
        <msup>
          <mi>t</mi>
          <mo>′</mo>
        </msup>
      </mstyle>
    </mrow>
    <annotation encoding="application/x-tex">{\displaystyle \int _{t-r}^{t}E(t')dt'}</annotation>
  </semantics>
</math></span><img src="../I/m/6d5dee5c41be26986ff2b8abf0c1fa2c196db968.svg" class="mwe-math-fallback-image-inline" aria-hidden="true" style="vertical-align: -2.505ex; width:12.375ex; height:6.343ex;"></span>
```

**Comment by Jaifroid** [comment]
Yes @kelson42, that would be a very good solution. I've tested that ZIM, and it would be easy for me to adapt my MathML detection to use the string(s) given in that element. I currently look for the alttext inside the math img element, but it would take me five minutes to adapt the code to use the math element instead, and it actually strikes me as a more elegant solution, because it does provide more semantic information about the TeX string.

**Comment by kelson42** [comment]
@ISNIT0 Could we easily copy the `alttex` attribute from one node to the other? That would avoid us to copy the whole node?

**Comment by kelson42** [comment]
We should really do something about this really old bug. I don't want to close the 1.9 effort with a clear answer about it.

**Comment by Jaifroid** [comment]
@kelson42 The option you give above about keeping the math node would work for me, if it's the simplest and quickest way to close this issue.
